### DISCUSSION TITLE
Fetching data too many times

### DISCUSSION THREAD

**Discussion by Gw3i** [neuefische/questions#117] [Web]
Hello everyone!
I'm currently fetching data from the [YouTube API](https://developers.google.com/youtube/v3/docs/search/list).
I've a limited amount of quotas per day. I just want to fetch the data once clicked on the search button.
Nonetheless the data is fetched before clicking on Search returning channelNames with the name "undefined" and is fetching a few more times, so my quotas are very fast spent.
So, my question is how just to fetch once after clicking on Search. 
I'm aware of using `useEffect`, but don't know where to put it. 
I'm also using a `useFetch`hook which is already using useEffect.

Here a screenshot of items fetched before searching:
<img width="812" alt="Screenshot 2022-08-29 at 10 06 11" src="https://user-images.githubusercontent.com/101993818/187154468-750ea664-3f3b-4074-acaa-0738b01a0feb.png">

The useFetch hook:
```javascript
import axios from 'axios';
import { useEffect, useState } from 'react';

export default function useFetch(url) {
	const [state, setState] = useState({
		data: null,
		loading: false,
		error: null,
	});

	useEffect(() => {
		setState(({ data }) => ({
			data,
			error: null,
			loading: true,
		}));
		axios
			.get(url)
			.then(({ data }) => {
				setState({
					error: null,
					data,
					loading: false,
				});
			})
			.catch(error => {
				setState(({ data }) => ({
					data,
					error,
					loading: false,
				}));
			});
	}, [url]);

	return state;
}
```

My code: 
```javascript
export default function SearchForm() {
	const {
		register,
		handleSubmit,
		reset,
		formState: { errors },
	} = useForm({ criteriaMode: 'all' });

	const [channelNames, setChannelNames] = useState(null);

	function onSearch(data) {
		setChannelNames(data);
		reset();
	}

	const fetchedChannels = `https://youtube.googleapis.com/youtube/v3/search?part=snippet&maxResults=3&q=${channelNames?.playlistSearch}&type=channel&key=${process.env.NEXT_PUBLIC_API_KEY}`;

	const { data } = useFetch(fetchedChannels);

	const channelItems = data?.items?.map(item => item.snippet);

	return (
		<>
			<StyledForm onSubmit={handleSubmit(onSearch)}>
				<StyledLabel htmlFor="videoPlaylist">
					Search for a YouTube channels playlist
					<input
						{...register('playlistSearch', {
							required: 'This field is required',
						})}
						placeholder="channel name..."
						name="playlistSearch"
						type="text"
						id="videoPlaylist"
					/>
					<ErrorMessage
						errors={errors}
						name="playlistSearch"
						render={({ messages }) =>
							messages &&
							Object.entries(messages).map(([type, message]) => (
								<StyledInputWarning key={type} role="alert">
									{message}
								</StyledInputWarning>
							))
						}
					/>
				</StyledLabel>
				<StyledButton>Search</StyledButton>
			</StyledForm>
			<ChannelsCard channelItems={channelItems} />
		</>
	);
}

```

My current Pull Request: https://github.com/Gw3i/capstone-project/pull/30
My App: https://capstone-project-cb7gmvycg-gw3i.vercel.app/create

Thank you for your time!

**Comment by mpagels** [comment]
`useEffect` runs, after the component was rendered the first time to the screen. And in your case, every time `url` is changed.
Everything inside `useEffect` is running. The `set` function and the `axios` function. 

If you want the `fetch` happening, only when you trigger a click, I would recommend putting the `axios` method inside a `handler` function, that is only triggered, when you clicked something.

Example

```js
// useState implementation (not shown here)

function myhandlerFunction() {
    axios.get(url).then(({ data }) => {setData(data)})
}

return <button onClick={myhandlerFunction}>Click me</button>
```

**Reply by Gw3i** [reply]
@mpagels Thanks for the answer!
I'm not really sure where to put this function. Should it be in my component?

**Reply by mpagels** [reply]
This function should be there, where it has access to the `set` function to save the data from the fetch response. If you need the information global, then put it in your store. If you need the information just in one component, then put it in this component and save the fetch response in a state.
